core/services/pipeline/orm.go

```diff
@@ -58,99 +55,126 @@ func (o *orm) CreateSpec(ctx context.Context, tx *gorm.DB, pipeline Pipeline, ma
+			return restart, err
+		}
 
-			taskRuns := []TaskRun{}
-			// Reload task runs, we want to check for any changes while the run was ongoing
-			if err = tx.Select(&taskRuns, `SELECT * FROM pipeline_task_runs WHERE pipeline_run_id = $1`, run.ID); err != nil {
-				return err
-			}
+		taskRuns := []TaskRun{}
+		// Reload task runs, we want to check for any changes while the run was ongoing
+		if err = sqlx.Select(tx, &taskRuns, `SELECT * FROM pipeline_task_runs WHERE pipeline_run_id = $1`, run.ID); err != nil {
+			return restart, err
+		}
 
-			// Construct a temporary run so we can use r.ByDotID
-			tempRun := Run{PipelineTaskRuns: taskRuns}
-
-			// Diff with current state, if updated, swap run.PipelineTaskRuns and early return with restart = true
-			for i, tr := range run.PipelineTaskRuns {
-				if !tr.IsPending() {
-					continue
-				}
-
-				// Look for new data
-				if taskRun := tempRun.ByDotID(tr.DotID); taskRun != nil {
-					// Swap in the latest state
-					run.PipelineTaskRuns[i] = *taskRun
-					restart = true
-				}
-			}
+		// Construct a temporary run so we can use r.ByDotID
+		tempRun := Run{PipelineTaskRuns: taskRuns}
 
-			if restart {
-				return nil
+		// Diff with current state, if updated, swap run.PipelineTaskRuns and early return with restart = true
+		for i, tr := range run.PipelineTaskRuns {
+			if !tr.IsPending() {
+				continue
 			}
 
-			// Suspend the run
-			run.State = RunStatusSuspended
-			if _, err = tx.NamedExec(`UPDATE pipeline_runs SET state = :state`, run); err != nil {
-				return err
-			}
-		} else {
-			// Simply finish the run, no need to do any sort of locking
-			if run.Outputs.Val == nil || len(run.Errors) == 0 {
-				return errors.Errorf("run must have both Outputs and Errors, got Outputs: %#v, Errors: %#v", run.Outputs.Val, run.Errors)
-			}
-			sql := `UPDATE pipeline_runs SET state = :state, finished_at = :finished_at, errors= :errors, outputs = :outputs WHERE id = :id`
-			if _, err = tx.NamedExec(sql, run); err != nil {
-				return err
+			// Look for new data
+			if taskRun := tempRun.ByDotID(tr.DotID); taskRun != nil && !taskRun.IsPending() {
+				// Swap in the latest state
+				run.PipelineTaskRuns[i] = *taskRun
+				restart = true
 			}
 		}
 
-		sql := `
+		if restart {
+			return restart, nil
+		}
+
+		// Suspend the run
+		run.State = RunStatusSuspended
+		if _, err = sqlx.NamedExec(tx, `UPDATE pipeline_runs SET state = :state`, run); err != nil {
+			return restart, err
+		}
+	} else {
+		// Simply finish the run, no need to do any sort of locking
+		if run.Outputs.Val == nil || len(run.Errors) == 0 {
+			return restart, errors.Errorf("run must have both Outputs and Errors, got Outputs: %#v, Errors: %#v", run.Outputs.Val, run.Errors)
+		}
+		sql := `UPDATE pipeline_runs SET state = :state, finished_at = :finished_at, errors= :errors, outputs = :outputs WHERE id = :id`
+		if _, err = sqlx.NamedExec(tx, sql, run); err != nil {
+			return restart, err
+		}
+	}
```

Comment:
after that `if restart {`, restart can never be true, so you could just have `return false, ...` in the next 3 returns. Also might be worth wrapping all 3 errors